command/agent/dns.go

```diff
@@ -877,6 +877,17 @@ func (d *DNSServer) handleRecurse(resp dns.ResponseWriter, req *dns.Msg) {
 
 // resolveCNAME is used to recursively resolve CNAME records
 func (d *DNSServer) resolveCNAME(name string) []dns.RR {
+	// If the CNAME record points to a Consul address, resolve it internally
+	if strings.HasSuffix(name, "."+strings.ToLower(d.agent.config.Domain)) {
```

Comment:
I think you'll need a `ToLower()` on name as well, since you don't know the incoming case from the user.